### PR TITLE
feat: Github Artifact Provider

### PR DESCRIPTION
**Pull request by HazAT**
aka. we can kill Zeus

giphy-4

You need to use Github actions to upload your artifacts, here is one example from the JS repo:

```yaml
  job_artifacts:
    name: Artifacts Upload
    needs: job_build
    runs-on: ubuntu-latest
    if: "contains(github.ref, 'release/')"
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v1
      - uses: actions/cache@v2
        with:
          path: |
            ${{ github.workspace }}/node_modules
            ${{ github.workspace }}/packages/**/node_modules
            ${{ github.workspace }}/packages/**/build
            ${{ github.workspace }}/packages/**/dist
            ${{ github.workspace }}/packages/**/esm
          key: ${{ runner.os }}-${{ github.sha }}
      - name: Pack
        run: yarn pack:changed
      - name: Archive Artifacts
        uses: actions/upload-artifact@v2
        with:
          name: ${{ github.sha }}
          path: |
            ${{ github.workspace }}/packages/browser/build/**
            ${{ github.workspace }}/packages/integrations/build/**
            ${{ github.workspace }}/packages/apm/build/**
            ${{ github.workspace }}/packages/tracing/build/**
            ${{ github.workspace }}/packages/**/*.tgz
```

the important part is this:

```yaml
- name: Archive Artifacts
        uses: actions/upload-artifact@v2
        with:
          name: ${{ github.sha }}
          path: |
            ${{ github.workspace }}/packages/browser/build/**
            ${{ github.workspace }}/packages/integrations/build/**
            ${{ github.workspace }}/packages/apm/build/**
            ${{ github.workspace }}/packages/tracing/build/**
            ${{ github.workspace }}/packages/**/*.tgz
```

using `actions/upload-artifact@v2` gha + setting the name to the `${{ github.sha }}` is the key since there is no other way to uniquely identify uploaded artifacts.

config in `craft.yaml`
```yaml
statusProvider:
  name: github
artifactProvider:
  name: github
```

### CODE REVIEW

File: src/artifact_providers/github.ts
Comment (tonyo) [inline src/artifact_providers/github.ts:80]: This returns a paginated response, right?
If yes, this might not work if the desired artifact is not on the first page.

File: src/artifact_providers/github.ts
Comment (tonyo) [inline src/artifact_providers/github.ts:96]: Curious if there's a chance we can have two or more artifacts with the same name?

File: src/artifact_providers/github.ts
Comment (HazAT) [reply src/artifact_providers/github.ts:80]: Correct, they are sorted by upload date but still, it could happen, so I will add paging.

File: src/artifact_providers/github.ts
Comment (HazAT) [reply src/artifact_providers/github.ts:96]: Yes if you have multiple jobs I guess it can happen, I will also make this more robust.

File: src/artifact_providers/github.ts
Comment (tonyo) [inline src/artifact_providers/github.ts:106]: Looks like this have a potential of entering the infinite loop if the condition holds initially, because both `total_count` and `per_page` don't change.

File: src/artifact_providers/github.ts
Comment (tonyo) [inline src/artifact_providers/github.ts:91]: Looking at the API, I think it also expects the page number

**Review by tonyo — CHANGES_REQUESTED** [review]

**Review by tonyo — APPROVED** [review]
LET'S DO THIS

**Comment by HazAT** [comment]
giphy-3

Will test before merging, only will merge it once confirmed the whole flow works.

**Pull request merged**
